chainer/graph_optimizations/static_graph_utilities.py

```diff
@@ -0,0 +1,197 @@
+import inspect
+
+import chainer
+
+
+def static_code(*dec_args, **dec_kwargs):
+    """Decorator to mark a function for inclusion in the static schedule.
+
+    This decorator is used to mark a function or method to be included
+    in the static schedule. This will only occur if the function is
+    called (directly or deeply) from inside a static chain's `__call__()`
+    method. That is, if a chain's `__call__()` uses the `@static_graph`
+    decorator, then any code that is executed while inside `__call__()`
+    that uses `@static_code` will be included in the corresponding
+    static schedule. Such code will be added to the static schedule in the
+    order that it was called.
+
+    This decorator should be applied to any code that needs to run each
+    iteration. This should ideally only include the code that performs
+    the actual forward and/or backward computations, and not include code
+    for initializaing parameters, checking types, etc.. If the user would
+    like to include any other code in a static chain's `__call__()` method
+    that needs to run every iteration, then it should also use this
+    decorator.
+
+    Usage:
+
+    This decorator can be applied to either a function or a method (typically
+    of a `FunctionNode`). There are no required arguments, and so a user can
+    apply it to "side effect" code to cause an operation to be executed each
+    iteration. The more usual use case is where the core framework code
+    will apply it to the all of (and only) the functions
+    that actually perform the computations needed to compute the forward
+    and backward passes.
+
+    The simplest usage is when we would like to force a particular
+    user-defined function to run each iteration. For example, such a function
+    might increment a counter, check conditions, and possibly print
+    information to the console. In this use, it is only required to add
+    this decorator to the function definition and then call it during
+    the first iteration from the context of the static chain's
+    `__call__()` method.
+
+    There are no constraints on the function's
+    arguments, unless the function needs to access, modify, and/or
+    create ndarrays that are used by other functions in the static
+    schedule. In these cases, the following contraints specify how
+    such arrays should be passed into and returned from a function
+    that uses this decorator.
```

Comment:
A short example of recommended usage would help here.

Reply:
Thanks for the suggestion. I agree, but I am trying to decide if it is better to include it here or to include in the documentation file (static_graph.rst).

Reply:
The example is in the "Usage" section:
https://github.com/chainer/chainer/pull/4811/files#diff-d27b1f920f51219d696c5a6f8847728fR28